### DISCUSSION TITLE
use setControllerReference

### DISCUSSION THREAD

**Discussion by Jerry-yz** [kubernetes-sigs/kubebuilder#3944] [Support] [answered]
when i use kubebuilder, controllerutil.SetControllerReference() relation crd and statefulset;
but crd is deleted, sts not delete
how can use it

**Comment by camilamacedo86** [comment, marked as answer]
Kubebuilder provides a plugin that scaffolds the codebase to manage and deploy an image on the cluster. For more details, see the [deploy image plugin documentation](https://kubebuilder.io/plugins/deploy-image-plugin-v1-alpha). By checking the testdata dir in the root directory of Kubebilder, you will find various examples. Then, examining the controller scaffolded within this plugin will provide code examples and links to the relevant documentation. See [this example](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/internal/controller/busybox_controller.go#L285-L303).

To understand the ownerRef, check: https://kubernetes.io/docs/tasks/administer-cluster/use-cascading-deletion/

By usingcontrollerutil.SetControllerReference() it will set that the CRD  ownerRef to the StatefulSet. So, I would suggest you check:
- How are you deleting. If the CRD is manually deleted without cascading deletion (e.g., using kubectl delete --cascade=orphan), the StatefulSet will not be deleted because the garbage collector is instructed not to delete dependent resources.
-  If the StatefulSet and the CRD are in different namespaces, I am guessing that the garbage collector might not delete the StatefulSet because it maybe do not recognize the owner reference across namespaces. (`This one is TBD / Required to confirm`)
- Verify your configuration  to ensure that the ownerRef is set

If the info above provide does not help you out, for we better help you out we would need to know:
- How are you using controllerutil.SetControllerReference()
- How are you deleting
- What is the cluster version / k8s

In other words, we would need to have more further details. Enough for example to reproduce the scenario. 
So, I am closing this one, but please feel free to open a support issue against this project if you need to.
